semgrep-core/src/engine/Selector.ml

```diff
@@ -0,0 +1,130 @@
+  | Not of sformula
+
+(*****************************************************************************)
+(* Selecting methods *)
+(*****************************************************************************)
+
+(* this will be adjusted later in range_to_pattern_match_adjusted *)
+let fake_rule_id (id, str) =
+  { PM.id = string_of_int id; pattern_string = str; message = "" }
+
+let match_selector (sel_opt : selector option) : AR.ranges =
+  let get_match (x : RP.times RP.match_result) = x.matches in
+  match sel_opt with
+  | None -> failwith "empty And; no positive terms in And"
+  | Some selector ->
+      List.map AR.match_result_to_range
+        (get_match (Lazy.force selector.lazy_matches))
+
+let select_from_ranges file (sel_opt : selector option) (ranges : AR.ranges) :
+    AR.ranges =
+  let pattern_match_from_binding selector (mvar, mval) =
+    {
+      PM.rule_id = fake_rule_id (selector.pid, selector.pstr);
+      PM.file;
+      PM.range_loc = Visitor_AST.range_of_any (M.mvalue_to_any mval);
+      PM.tokens = lazy (M.ii_of_mval mval);
+      PM.env = [ (mvar, mval) ];
+    }
+  in
+  let select_from_range range =
+    match sel_opt with
+    | None -> [ range ]
+    | Some selector -> (
+        match
+          List.find_opt
+            (fun (mvar, _mval) -> M.equal_mvar selector.mvar mvar)
+            range.AR.mvars
+        with
+        | None -> []
+        | Some binding ->
+            (* make a pattern match, then use AR.match_result_to_range *)
+            [
+              AR.match_result_to_range
+                (pattern_match_from_binding selector binding);
+            ] )
+  in
+  List.flatten (List.map select_from_range ranges)
+
+let selector_equal s1 s2 = s1.mvar = s2.mvar
+
+(*****************************************************************************)
+(* Converter *)
+(*****************************************************************************)
+
+let formula_to_sformula match_func formula =
+  let rec formula_to_sformula formula =
+    let selector_from_formula f =
+      match f with
+      | R.Leaf (R.P ({ pat = Sem (pattern, _); pid; pstr }, None)) -> (
+          match pattern with
+          | G.E (G.N (G.Id ((mvar, _), _))) ->
+              Some
+                {
+                  mvar;
+                  pid;
+                  pstr;
+                  lazy_matches = lazy (match_func [ (pattern, pid, pstr) ]);
+                }
+          | _ -> None )
+      | _ -> None
+    in
+    let rec remove_selectors (selector, acc) formulas =
+      match formulas with
+      | [] -> (selector, acc)
+      | x :: xs ->
+          let selector, acc =
+            match (selector, selector_from_formula x) with
+            | None, None -> (None, x :: acc)
+            | Some s, None -> (Some s, x :: acc)
+            | None, Some s -> (Some s, acc)
+            | Some s1, Some s2 ->
+                if selector_equal s1 s2 then (Some s1, acc) else (None, [])
+            (* This will never be valid *)
```

Comment:
What do you mean by valid here? Do you mean that the pattern-and is not valid?

Maybe I'm missing something important here, but it seems to me that:
```
patterns:
  - pattern: $X
  - pattern: $Y
```
which should match something will become `(None, [])` which matches nothing?